### PR TITLE
Meets #7515: Copy/Move does not update available statuses

### PR DESCRIPTION
**Pull request by slawa85**
[`* `#7515` Copy/Move does not update available statuses`](https://www.openproject.org/work_packages/7515)

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage. <a href='http://netfighter-feature-7515-copy-or-move-does-not-update-available-.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

File: app/controllers/api/experimental/work_packages_controller.rb [hunk 1]
```diff
@@ -35,9 +35,9 @@ class WorkPackagesController < ApplicationController
       DEFAULT_SORT_ORDER = ['parent', 'desc']
 
       include ApiController
-      include Concerns::GrapeRouting
-      include Concerns::ColumnData
-      include Concerns::QueryLoading
+      include Api::Experimental::Concerns::GrapeRouting
+      include Api::Experimental::Concerns::ColumnData
+      include Api::Experimental::Concerns::QueryLoading
```
Comment (ulferts) [inline app/controllers/api/experimental/work_packages_controller.rb:40]: :+1:

File: app/controllers/work_packages/moves_controller.rb [hunk 2]
```diff
@@ -30,7 +30,7 @@
 class WorkPackages::MovesController < ApplicationController
   default_search_scope :work_packages
   before_filter :find_work_packages, :check_project_uniqueness
-  before_filter :authorize
+  before_filter :authorize, except: [:change_type]
```
Comment (ulferts) [inline app/controllers/work_packages/moves_controller.rb]: this here is dangerous, the action needs to be protected.

File: app/controllers/work_packages/moves_controller.rb [hunk 3]
```diff
@@ -97,6 +97,19 @@ def default_breadcrumb
     l(:label_move_work_package)
   end
 
+  def change_type
```
Comment (ulferts) [inline app/controllers/work_packages/moves_controller.rb]: Why aren't you reusing the #new action and alter it to also work correctly with the types? AFAIK, when updating the project, it is already called, so it makes more sense to me to also reuse it for updating a type. Maybe the updating the statuses functionality can be moved completely to the prepare_for_work_package_move function.

File: app/controllers/work_packages/moves_controller.rb [hunk 4]
```diff
@@ -97,6 +97,19 @@ def default_breadcrumb
...
+    prepare_for_work_package_move
+    wp_type = @target_project.types.find_by_id(params[:new_type_id])
+    @statuses = []
+    @work_packages.each do |wp|
+      wp_status = wp.status
+      current_user_role = current_user.roles_for_project(@target_project)
+      @statuses += wp_status.find_new_statuses_allowed_to(current_user_role, wp_type)
+    end
+    @statuses.uniq!
+    render layout: false
```
Comment (ulferts) [inline app/controllers/work_packages/moves_controller.rb]: When switching the type, the user no longer has the option to not change the status. He has that for the other attributes. Any reasons for this?

File: config/routes.rb [hunk 5]
```diff
@@ -416,9 +416,14 @@
 
     # move bulk of wps
     get 'move/new' => 'work_packages/moves#new', :on => :collection, :as => 'new_move'
+    get 'move/change_type' => 'work_packages/moves#change_type', :on => :collection
```
Comment (ulferts) [inline config/routes.rb]: While there are already routes defined like this, I would rather you add a

<pre>
collection do
  get :change_type
end
</pre>

to the block with the member action as well.

File: config/routes.rb [hunk 6]
```diff
@@ -416,9 +416,14 @@
...
     post 'move' => 'work_packages/moves#create', :on => :collection, :as => 'move'
     # move individual wp
-    resource :move, :controller => 'work_packages/moves', :only => [:new, :create]
+    resource :move, :controller => 'work_packages/moves', only: [:new, :create] do
+      member do
+        get :change_type
```
Comment (ulferts) [inline config/routes.rb]: Please add route specs (`spec/routing`) for the added routes.

File: app/controllers/work_packages/moves_controller.rb (hunk 3 above)
Comment (slawa85) [inline app/controllers/work_packages/moves_controller.rb]: I use this method in other ticket also, thats why I prefered to create a separate action for change status

**Comment by apfelfabrik** [comment]
wow, @houndci is noisy!

**Comment by myabc** [comment]
btw #1944 will quieten Hound's complaints about multi-line blocks.

**Comment by slawa85** [comment]
This PR should be merged after this PR https://github.com/opf/openproject/pull/1974

**Pull request closed without merging**